Python: Add support for handling numpy vectors and JIT compilation with Numba (Closes #34)

dmey opened the pull request:
This PR adds the following improvements:

1. Support for handling numpy vectors as inputs/outputs and greater run time performance using Numba: when Numba is found, all functions are compiled using the Numba just in time (JIT) compiler. Note that there is a small overhead the first time the code is run due to the JIT.

2. CI: jobs are now split into 1-2: Python (with/without Numba); 3: C#; 4: R

3. README: content and style.

C-H-Simpson commented:
An unfortunate effect of wrapping in this way is that the docstrings are hidden by the wrapper, so e.g. help(psy.GetRelHumFromVapPres) gives "Help on DUFunc in module numba.npyufunc.dufunc object...". There may be a way of fix this, I don't know.

dmey commented:
> An unfortunate effect of wrapping in this way is that the docstrings are hidden by the wrapper, so e.g. help(psy.GetRelHumFromVapPres) gives "Help on DUFunc in module numba.npyufunc.dufunc object...". There may be a way of fix this, I don't know.

That's right -- [it looks](https://stackoverflow.com/a/29580585) like this will be a limitation so will should probably ask people to just consult the [online API](https://psychrometrics.github.io/psychrolib/api_docs.html) -- not great but given the other options this seems reasonable.

dmey commented:
@C-H-Simpson how did you handle return type for the Calc functions (e.g. `CalcPsychrometricsFromRelHum`)? I may need to change those for Numba...

C-H-Simpson commented:
@dmey The Calc... functions can't be straightforwardly vectorized because they return a tuple. 
They should be excluded from the list of functions to wrap with _vectorized_. As the functions it calls to do the calculations are vectorized, it will return a tuple of vectors.
I have tested this and it works.
I don't think there would be much of a speed-up from wrapping the Calc... functions in numba anyway, as they don't actually do any calculations themselves.

So I would recommend just not vectorizing them.

dmey commented:
@C-H-Simpson thanks -- this makes perfect sense! I will just need to check a few more things/tests so I will need a bit longer before we can release a new version -- hopefully will not be too long though!

dmey commented:
@didierthevenard this PR is now ready to be merged in -- all tests pass. Please review when you have a moment and let me know for any issue.

didierthevenard commented on .travis.yml:
```diff
@@ -1,46 +1,85 @@
+  local_dir: build/html
+  on:
+    branch: master
+    condition: $USE_NUMBA == "no"
```
Add newline at end of file

didierthevenard commented on README.md:
```diff
@@ -1,32 +1,46 @@
 ## Documentation
 
-The API documentation is available [here](https://psychrometrics.github.io/psychrolib/api_docs.html). Please note that although the API describes the Python version of the library, the API is common across all the supported language implementations. In R please note that (1) constants, like `ZERO_FAHRENHEIT_AS_RANKINE` are not exported (i.e. not directly accessible to users), (2) functions accept ...
+Please see the [Python API documentation](https://psychrometrics.github.io/psychrolib/api_docs.html) for the common API across all the supported language implementations. In Python, for best runtime performance and array support, please install [Numba](https://numba.pydata.org/). In R (1) constants, like `ZERO_FAHRENHEIT_AS_RANKINE` are not exported (i.e. not directly accessible to users), (2) fu ...
```
Should mention that installation of numba is optional, that the library stands on its own without numba. Maybe suggest installing numba only for array support.

didierthevenard commented on README.md:
```diff
@@ -1,32 +1,46 @@
+## Installation
 
-- Python: from the [Python Package Index (PyPI)](https://pypi.org/project/PsychroLib/).
+- Python: from the [Python Package Index (PyPI)](https://pypi.org/project/PsychroLib/). For best runtime performance and array support, install [Numba](https://numba.pydata.org/).
```
Same comment as before regarding numba.

didierthevenard commented on src/python/psychrolib.py:
```diff
@@ -172,6 +172,13 @@ def SetUnitSystem(Units: UnitSystem) -> None:
     else:
         PSYCHROLIB_TOLERANCE = 0.001
 
+    if has_numba:
```
We need comments before this line, to explain to users why this section is needed and what it does.

didierthevenard commented on src/python/psychrolib.py:
```diff
@@ -1459,3 +1466,20 @@ def CalcPsychrometricsFromRelHum(TDryBulb: float, RelHum: float, Pressure: float
+            func_list.append(func)
+except ImportError:
+    has_numba = False
+    vectorize = None
```
Newline needed at end of file.

didierthevenard commented on src/python/psychrolib.py:
```diff
@@ -1459,3 +1466,20 @@ def CalcPsychrometricsFromRelHum(TDryBulb: float, RelHum: float, Pressure: float
     DegreeOfSaturation = GetDegreeOfSaturation(TDryBulb, HumRatio, Pressure)
     return HumRatio, TWetBulb, TDewPoint, VapPres, MoistAirEnthalpy, MoistAirVolume, DegreeOfSaturation
+
+
```
Comments are needed to explain what this section does (and also why some functions are excluded.

dmey commented:
> A few comments added to the proposed changes. No issue with the code itself, it's mostly that new code that has been added (around the use of numba and vectorization) needs to have explaining comments.

👍. Please see now.

didierthevenard commented on src/python/psychrolib.py:
```diff
@@ -1459,3 +1466,24 @@ def CalcPsychrometricsFromRelHum(TDryBulb: float, RelHum: float, Pressure: float
+            func_list.append(func)
+except ImportError:
+    has_numba = False
+    vectorize = None
```
What's the purpose of this line? I have the impression that it doesn't do anything.

dmey replied:
that's right -- dead code I forgot to remove from the previous version. 👍

dmey commented:
> Thanks for the added comments. I fixed one typo and added a general comment about the last section in the code (the entire chunk that deals with Numba) - feel free to change if you can think of a better wording. Also wondering about the usefulness of the last line in the code.

Thanks -- looks good. Unused var removed in https://github.com/psychrometrics/psychrolib/pull/57/commits/af6440655f621a38365881ba1f951674d377aa74.
